src/Components/TestExplorer.fs

```diff
@@ -131,59 +131,91 @@ module DotnetTest =
                         xpathSelector.SelectString
                             $"/t:TestRun/t:TestDefinitions/t:UnitTest[{idx}]/t:TestMethod/@className"
 
+                    // This code assumes there will only
                     let test =
                         xpathSelector.SelectString $"/t:TestRun/t:TestDefinitions/t:UnitTest[{idx}]/t:TestMethod/@name"
 
-                    $"{className}.{test}", executionId)
-                |> Map.ofArray
+                    $"{className}.{test}", test, executionId)
 
-            let tests =
+            let unmappedTests, mappedTests =
                 projectWithTests.Tests
-                |> Array.map (fun t ->
-                    // If the test contains a single quote, xpath expressions cannot escape this and it leads to a world of pain.
-                    let executionId = Map.find t.FullName testDefinitions
+                |> Array.sortByDescending (fun t -> t.FullName)
+                |> Array.fold
+                    (fun (tests, mappedTests) (t) ->
+                        let linkedTests, remainingTests =
+                            tests
+                            |> Array.partition (fun (fullName: string, _, _) -> fullName.StartsWith t.FullName)
 
-                    let outcome =
-                        xpathSelector.SelectString
-                            $"/t:TestRun/t:Results/t:UnitTestResult[@executionId='{executionId}']/@outcome"
+                        if Array.isEmpty linkedTests then
+                            remainingTests, mappedTests
+                        else
+                            remainingTests, ([| yield! mappedTests; (t, linkedTests) |]))
+                    (testDefinitions, [||])
 
-                    let errorInfoMessage =
-                        xpathSelector.TrySelectString
-                            $"/t:TestRun/t:Results/t:UnitTestResult[@executionId='{executionId}']/t:Output/t:ErrorInfo/t:Message"
+            let unmappedTestNames =
+                unmappedTests |> Array.map (fun (fullName, _, _) -> fullName)
 
-                    let errorStackTrace =
-                        xpathSelector.TrySelectString
-                            $"/t:TestRun/t:Results/t:UnitTestResult[@executionId='{executionId}']/t:Output/t:ErrorInfo/t:StackTrace"
+            logger.Debug("Unmapped tests", unmappedTestNames)
```

Comment:
it looks like we do this second array walk only to emit the names of the unmapped tests - do you see this happening anywhere, or was this a debugging aid while you were developing? if the latter, can you delete it so we do less work?

Reply:
Not that I saw this happening, but I'm pretty sure this implementation is a plaster over the issue rather than the right fix. The intention was to add it in to make it easier to review which tests don't get correctly mapped in the implementation, but given it's a few lines and walks over the structure, I'll remove it.

Reply:
https://github.com/ionide/ionide-vscode-fsharp/issues/1756#issuecomment-1445485152 This example I don't think is fixed by this implementation, but maybe that could be addressed in a subsequent PR.

Reply:
I agree - happy to take incremental improvements to this as they come!